## Machine Learning Pipeline

As the class practice, the students will be required to develop a machine learning pipeline using the `Churn_Modelling_train_test.csv` dataset.

**About dataset**

This dataset is obained from [kaggle](https://www.kaggle.com/datasets/shubhammeshram579/bank-customer-churn-prediction?resource=download). It contains information on bank customers who either left the bank or continue to be a customer. The dataset includes the following attributes:

* Customer ID: A unique identifier for each customer
* Surname: The customer's surname or last name
* Credit Score: A numerical value representing the customer's credit score
* Geography: The country where the customer resides (France, Spain or Germany)
* Gender: The customer's gender (Male or Female)
* Age: The customer's age.
* Tenure: The number of years the customer has been with the bank
* Balance: The customer's account balance
* NumOfProducts: The number of bank products the customer uses (e.g., savings account, credit card)
* HasCrCard: Whether the customer has a credit card (1 = yes, 0 = no)
* IsActiveMember: Whether the customer is an active member (1 = yes, 0 = no)
* EstimatedSalary: The estimated salary of the customer
* Exited: Whether the customer has churned (1 = yes, 0 = no)


### Exploratory Data Analysis

In this section the students are required to do an EDA to understand the dataset.

In [1]:
# import packages
...
# import libraries
import pandas as pd
import joblib
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import OneHotEncoder
from sklearn.metrics import f1_score

In [2]:
# Load the dataset
df = pd.read_csv(r"C:\Users\ASUS\Desktop\EADA\3º trimestre\ML Operations\Activities\mlops-and-system-design\session 2\Ejercicios\Churn_Modelling_train_test.csv")

FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\ASUS\\Desktop\\EADA\\3º trimestre\\ML Operations\\Activities\\mlops-and-system-design\\session 2\\Ejercicios\\Churn_Modelling_train_test.csv'

In [ ]:
df.info

<bound method DataFrame.info of       RowNumber  CustomerId        Surname  CreditScore Geography  Gender  \
0          4784    15729224       Jennings          710    France  Female   
1          1497    15799156     Okwuadigbo          569     Spain    Male   
2          1958    15674922        Beavers          710    France    Male   
3          9174    15653572       Thornton          673     Spain    Male   
4          9748    15775761  Iweobiegbunam          610   Germany  Female   
...         ...         ...            ...          ...       ...     ...   
8996       5735    15596647      Henderson          768    France    Male   
8997       5192    15681075    Chukwualuka          682    France  Female   
8998       5391    15573851      Macrossan          735    France  Female   
8999        861    15807663       McGregor          667    France    Male   
9000       7271    15706268          Smith          697   Germany    Male   

       Age  Tenure    Balance  NumOfProduct

In [ ]:
# dataset analysis
df.head()



,RowNumber,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited
0,4784,15729224,Jennings,710,France,Female,37.0,5,0.00,2,1.0,0.0,115403.31,0
1,1497,15799156,Okwuadigbo,569,Spain,Male,38.0,8,0.00,2,0.0,0.0,79618.79,0
2,1958,15674922,Beavers,710,France,Male,54.0,6,171137.62,1,1.0,1.0,167023.95,1
3,9174,15653572,Thornton,673,Spain,Male,43.0,8,127132.96,1,0.0,1.0,6009.27,1
4,9748,15775761,Iweobiegbunam,610,Germany,Female,69.0,5,86038.21,3,0.0,0.0,192743.06,1


In [ ]:
df.shape

(9001, 14)

In [ ]:
# review categorical variables
print("value coutns",df['Geography'].value_counts())
print("mean",df['Geography'].mode())
print("%",df['Geography'].value_counts(normalize=True) * 100)

value coutns Geography
France     4505
Spain      2250
Germany    2245
Name: count, dtype: int64
mean 0    France
Name: Geography, dtype: object
% Geography
France     50.055556
Spain      25.000000
Germany    24.944444
Name: proportion, dtype: float64


In [ ]:
print("value coutns",df['Gender'].value_counts())
print("mean",df['Gender'].mode())
print("%",df['Gender'].value_counts(normalize=True) * 100)

value coutns Gender
Male      4906
Female    4095
Name: count, dtype: int64
mean 0    Male
Name: Gender, dtype: object
% Gender
Male      54.505055
Female    45.494945
Name: proportion, dtype: float64


In [ ]:
# review binary variables
...
print("value conts",df['Exited'].value_counts())
print("%",df['Exited'].value_counts(normalize=True) * 100)


value conts Exited
0    7161
1    1840
Name: count, dtype: int64
% Exited
0    79.557827
1    20.442173
Name: proportion, dtype: float64


In [ ]:
print("value conts",df['HasCrCard'].value_counts())
print("%",df['HasCrCard'].value_counts(normalize=True) * 100)

value conts HasCrCard
1.0    6351
0.0    2649
Name: count, dtype: int64
% HasCrCard
1.0    70.566667
0.0    29.433333
Name: proportion, dtype: float64


In [ ]:
print("value conts",df['IsActiveMember'].value_counts())
print("%",df['IsActiveMember'].value_counts(normalize=True) * 100)

value conts IsActiveMember
1.0    4633
0.0    4368
Name: count, dtype: int64
% IsActiveMember
1.0    51.472059
0.0    48.527941
Name: proportion, dtype: float64


In [ ]:
# review numerical variables
...
print("Mean", df['Balance'].mean())

print("Median", df['Balance'].median())

print("Min", df['Balance'].min())

print("Max", df['Balance'].max())

print("Std", df['Balance'].std())


Mean 76222.21082657482
Median 96997.09
Min 0.0
Max 250898.09
Std 62432.19815106162


In [ ]:
print("Mean", df['CreditScore'].mean())

print("Median", df['CreditScore'].median())

print("Min", df['CreditScore'].min())

print("Max", df['CreditScore'].max())

print("Std", df['CreditScore'].std())

Mean 650.681368736807
Median 652.0
Min 350
Max 850
Std 96.53959128284978


In [ ]:
print("Mean", df['Age'].mean())

print("Median", df['Age'].median())

print("Min", df['Age'].min())

print("Max", df['Age'].max())

print("Std", df['Age'].std())

Mean 38.90178111111111
Median 37.0
Min 18.0
Max 92.0
Std 10.450759502027477


In [ ]:
print("Mean", df['Tenure'].mean())

print("Median", df['Tenure'].median())

print("Min", df['Tenure'].min())

print("Max", df['Tenure'].max())

print("Std", df['Tenure'].std())


Mean 5.007888012443062
Median 5.0
Min 0
Max 10
Std 2.8940252619475744


In [ ]:
print("Mean", df['NumOfProducts'].mean())

print("Median", df['NumOfProducts'].median())

print("Min", df['NumOfProducts'].min())

print("Max", df['NumOfProducts'].max())

print("Std", df['NumOfProducts'].std())  

Mean 1.5314965003888457
Median 1.0
Min 1
Max 4
Std 0.5793982263367707


In [ ]:
print("Mean", df['EstimatedSalary'].mean())

print("Median", df['EstimatedSalary'].median())

print("Min", df['EstimatedSalary'].min())

print("Max", df['EstimatedSalary'].max())

print("Std", df['EstimatedSalary'].std()) 

Mean 100180.82396733695
Median 100556.98
Min 11.58
Max 199992.48
Std 57561.18953385744


In [ ]:
cols = [
    'Balance',
    'CreditScore',
    'Age',
    'Tenure',
    'NumOfProducts',
    'EstimatedSalary'
]

for col in cols:

    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)

    IQR = Q3 - Q1

    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR

    outliers = df[(df[col] < lower) | (df[col] > upper)]

    print(f"{col}: {outliers.shape[0]} outliers")

Balance: 0 outliers
CreditScore: 13 outliers
Age: 318 outliers
Tenure: 0 outliers
NumOfProducts: 50 outliers
EstimatedSalary: 0 outliers


**EDA Conclusions**

From this Exploratory Data Analysis we can retrieve the following information:
The data doenst contain NaN values. If we focus oin the numerical values we see that we have a lot of outliers of age but is not something relevant for this kidn of situation. for credit score and numebr of salaries we have 13 and 50 outliers but of 9001 is 3.5% for the numebr of products that is only a 0.55% of the data in outliers so its very little. The outliers are not significant enough to distort the model, so we can keep them without special treatment. Categorical variables like Geography and Gender should be encoded, and numerical variables may benefit from scaling

### ML Pipeline

In this section the students are required to create a ML pipeline to predict whether the customer left the bank or not.

In [ ]:
# import packages
import pandas as pd
!python -m pip install mlflow
import mlflow
from mlflow.models import infer_signature


[notice] A new release of pip is available: 24.3.1 -> 26.1.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [ ]:
class Transformer:
    def __init__(self):
        self.DROP_COLUMNS = ['CustomerId', 'Surname']
        self.CATEGORICAL_COLS = ['Geography', 'Gender']

        self.BINARY_FEATURES = ['HasCrCard', 'IsActiveMember']
    
    def transform(self, df: pd.DataFrame) -> pd.DataFrame:
        df = df.drop(columns=self.DROP_COLUMNS, errors='ignore')
        df = self._ensure_binary_int(df)

        df = pd.get_dummies(df, columns=self.CATEGORICAL_COLS, drop_first=False)
        return df
    
    def _ensure_binary_int(self, df: pd.DataFrame) -> pd.DataFrame:
        for col in self.BINARY_FEATURES:
            if col in df.columns:
                df[col] = df[col].astype(int)
        return df

In [ ]:

from sklearn.model_selection import train_test_split

X = df.drop('Exited', axis=1)
y = df['Exited']

X_train, X_test, y_train, y_test = train_test_split(
    X, y, 
    test_size=0.2, 
    random_state=42,
    stratify=y  
)

train_df = X_train.copy()
train_df['Exited'] = y_train

In [ ]:
from sklearn.model_selection import train_test_split

X = df.drop('Exited', axis=1)
y = df['Exited']

X_train, X_test, y_train, y_test = train_test_split(
    X, y, 
    test_size=0.2, 
    random_state=42,
    stratify=y  
)

train_df = X_train.copy()
train_df['Exited'] = y_train

def balance_dataset(df: pd.DataFrame) -> pd.DataFrame:
    df_y0 = df[df['Exited'] == 0]   
    df_y1 = df[df['Exited'] == 1]  
    min_size = len(df_y1)
    df_y0_balanced = df_y0.sample(n=min_size, random_state=42)
    df_balanced = pd.concat([df_y0_balanced, df_y1])
    df_balanced = df_balanced.sample(frac=1, random_state=42).reset_index(drop=True)
    return df_balanced

train_balanced = balance_dataset(train_df)

X_train_bal = train_balanced.drop('Exited', axis=1)
y_train_bal = train_balanced['Exited']

print(f"Train original: {len(train_df)}")
print(f"Train balanceado: {len(train_balanced)}")
print(f"Test sin balancear: {len(X_test)}")

Train original: 7200
Train balanceado: 2944
Test sin balancear: 1801


#### Apply Transformer to train and test sets

We apply the `Transformer` class defined above to both the balanced train set and the test set. We also fill NaN values using the median computed from the training set only (to avoid data leakage).

In [ ]:
transformer = Transformer()

X_train_transformed = transformer.transform(X_train_bal)
X_test_transformed = transformer.transform(X_test)

X_test_transformed = X_test_transformed.reindex(columns=X_train_transformed.columns, fill_value=0)

medians = X_train_transformed.median()
X_train_transformed = X_train_transformed.fillna(medians)
X_test_transformed = X_test_transformed.fillna(medians)

print("Shape X_train:", X_train_transformed.shape)
print("Shape X_test:", X_test_transformed.shape)
print("NaN in train:", X_train_transformed.isnull().sum().sum())
print("NaN in test:", X_test_transformed.isnull().sum().sum())

Shape X_train: (2944, 14)
Shape X_test: (1801, 14)
NaN in train: 0
NaN in test: 0


#### Train the Logistic Regression model

In [ ]:
# Create and train the model
model = LogisticRegression(max_iter=1000, random_state=42)
model.fit(X_train_transformed, y_train_bal)

print("Model trained successfully")

Model trained successfully


c:\Users\ASUS\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\linear_model\_logistic.py:406: ConvergenceWarning: lbfgs failed to converge after 1000 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=1000).
You might also want to scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


#### Make predictions on the test set

In [ ]:
# Predict on the test set
y_pred = model.predict(X_test_transformed)
print("First 10 predictions:", y_pred[:10])

First 10 predictions: [1 0 0 1 0 0 1 0 1 1]


#### Evaluate the model

In [ ]:
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

print(f"Accuracy on test: {accuracy:.4f}")
print(f"F1-score on test: {f1:.4f}")

Accuracy on test: 0.7068
F1-score on test: 0.4971


### MLFlow Integration

In this section the students are required to track the model and the parameters into MLFlow.

In [ ]:
# import packages for mlfow
import mlflow

Start a local Tracking server on port 8080 - it is necessary to import mlflow and open a new terminal

In [ ]:

mlflow.set_tracking_uri(uri="http://127.0.0.1:8080")

Visit the MLFlow UI on `http://127.0.0.1:8080`

In [ ]:
# Create a new MLflow Experiment
mlflow.set_experiment("Practice Experiment - Ahernandez")

<Experiment: artifact_location='mlflow-artifacts:/110626133455086486', creation_time=1779140218026, experiment_id='110626133455086486', last_update_time=1779140218026, lifecycle_stage='active', name='Practice Experiment - Ahernandez', tags={}, trace_location=None, workspace='default'>

In [ ]:
# Define hyperparameters used
params = {
    "model_type": "LogisticRegression",
    "max_iter": 1000,
    "random_state": 42,
    "balanced_train": True,
    "imputation": "median",
}

# Start an MLflow run
with mlflow.start_run():
    mlflow.log_params(params)
    mlflow.log_metric("accuracy", accuracy)
    mlflow.log_metric("f1_score", f1)
    mlflow.set_tag("Training Info", "Logistic Regression on Churn dataset (balanced train)")
    print("MLflow run logged successfully (without model artifact)")

MLflow run logged successfully (without model artifact)
🏃 View run receptive-stag-732 at: http://127.0.0.1:8080/#/experiments/110626133455086486/runs/66002401717442cf8b28bbd1e1da6e14
🧪 View experiment at: http://127.0.0.1:8080/#/experiments/110626133455086486


### Keep practicing

In [ ]:

from sklearn.ensemble import RandomForestClassifier

rf_params = {
    "model_type": "RandomForestClassifier",
    "n_estimators": 200,
    "max_depth": 10,
    "random_state": 42,
}

rf_model = RandomForestClassifier(
    n_estimators=rf_params["n_estimators"],
    max_depth=rf_params["max_depth"],
    random_state=rf_params["random_state"],
)
rf_model.fit(X_train_transformed, y_train_bal)

y_pred_rf = rf_model.predict(X_test_transformed)
accuracy_rf = accuracy_score(y_test, y_pred_rf)
f1_rf = f1_score(y_test, y_pred_rf)

print(f"Random Forest - Accuracy: {accuracy_rf:.4f}")
print(f"Random Forest - F1-score: {f1_rf:.4f}")

with mlflow.start_run():
    mlflow.log_params(rf_params)
    mlflow.log_metric("accuracy", accuracy_rf)
    mlflow.log_metric("f1_score", f1_rf)
    mlflow.set_tag("Training Info", "Random Forest on Churn dataset (balanced train)")
    print("Random Forest run logged in MLflow (params and metrics only)")

Random Forest - Accuracy: 0.7812
Random Forest - F1-score: 0.5817
Random Forest run logged in MLflow (params and metrics only)
🏃 View run languid-elk-683 at: http://127.0.0.1:8080/#/experiments/110626133455086486/runs/12f25baf1718423599c69307e26061bd
🧪 View experiment at: http://127.0.0.1:8080/#/experiments/110626133455086486
